File: src/SamuraiServer/SamuraiServer.Data/Models/Map.cs
using System;
using System.Collections.Generic;
using System.Linq;
using SamuraiServer.Data.Tiles;

namespace SamuraiServer.Data
{
    public class Map
    {
        public Map()
        {
            Id = Guid.NewGuid();
        }

        public Guid Id { get; private set; }
        public string Name { get; set; }
        public string ImageResource { get; set; }

        public int MinPlayers { get; set; }
        public int MaxPlayers { get; set; }
        public Dictionary<int, List<Unit>> StartingUnits { get; set; }

        public TileType[][] Tiles { get; set; }

        public string[] GetStringRepresentation()
        {
            return Tiles.Select(t => new string(t.Select(t2 => t2.StringRepresentation).ToArray())).ToArray();
        }

        public static Map FromStringRepresentation(Guid id, string[] rows)
        {
            Map m = new Map();
            m.Id = id;
            m.Tiles = new TileType[rows[0].Length][];
            for (int x = 0; x < m.Tiles.Length; x++)
            {
                m.Tiles[x] = new TileType[rows.Length];
                for (int y = 0; y < rows.Length; y++)
                {
                    TileType t;
                    switch (rows[y][x])
                    {

                        case '.':
                            t = new Grass();
                            break;

                        case '@':
                            t = new Rock();
                            break;

                        case 'T':
                            t = new Tree();
                            break;

                        case '~':
                            t = new Water();
                            break;

                        default:
                            continue;
                    }
                    m.Tiles[x][y] = t;
                }
            }
            return m;
        }
    }
}

File: src/Samurai.Client.JS/models/map.js
//using SamuraiServer.Data.Tiles;
define(['tiles'], function (Tiles) {

  function Map(opt) {
  //  public Map()
  //  {
  //      Id = Guid.NewGuid();
  //  }

    this.initialize(opt);
  }

  Map.prototype.toString = function () {
  //  return Tiles.Select(t => new string(t.Select(t2 => t2.StringRepresentation).ToArray())).ToArray();
  };

  Map.prototype.stringRepresentation = function () {
    var strMap = [],
        r = 0, c = 0;

    for( ; r < this.Tiles.length; r++) {
      strMap[r] = [];
      for(c = 0; c < this.Tiles[0].length; c++) {
        strMap[r][c] = this.Tiles[r][c].toString();
      }
    }

    return strMap;
  };

  Map.prototype.initialize = function (opt) {
    var key = null,
    default_args = {
  //      public Guid Id { get; private set; }
      Id : '',
      Name : '',
      ImageResource : '',
      MinPlayers : 0,
      MaxPlayers : 0,
  //      public Dictionary<int, List<Unit>> StartingUnits { get; set; }
      StartingUnits : {},
  //      public TileType[][] Tiles
      Tiles : []
  };

  opt || (opt = default_args);

  for(key in default_args) {
    if(typeof opt[key] == "undefined") opt[key] = default_args[key];
  }

  // opt[] has all the data - user provided and optional.
    for(key in opt) {
      this[key] = opt[key];
    }
  };

  Map.prototype.fromStringRepresentation = function fromString(id, rows) {
    var m = new Map({
          Id : id,
          Tiles : new Array(rows.length)
        }),
        x = 0, y = 0,
        t;          //TileType t;

    for( ; x < m.Tiles.length; x++) {
      m.Tiles[x] = new Array(rows[0].length);
        for(y = 0; y < rows[0].length; y++) {
        switch(rows[x][y]) {
          case '.':
              t = new Tiles.Grass();
              break;

          case '@':
              t = new Tiles.Rock();
              break;

          case 'T':
              t = new Tiles.Tree();
              break;

          case '~':
                t = new Tiles.Water();
                break;

            default:
                continue;
          }

          m.Tiles[x][y] = t;
        }
      }

    return m;
  };

  return Map;
});